packages/big-design/src/components/Alert/spec.tsx

```diff
@@ -1,97 +1,109 @@
 
+  expect(alert).toMatchSnapshot();
   expect(link).toBeInTheDocument();
   expect(link.href).toBe('http://localhost/#');
 });
 
-test('renders with external link', () => {
-  const { queryByRole, container } = render(
+test('renders with external link', async () => {
+  render(
     <Alert messages={[{ text: 'Success', link: { text: 'Link', href: '#', external: true, target: '_blank' } }]} />,
   );
 
-  expect(container.firstChild).toMatchSnapshot();
-
-  const link = queryByRole('link') as HTMLAnchorElement;
+  const alert = await screen.findByRole('alert');
+  const link = screen.getByRole('link') as HTMLAnchorElement;
 
+  expect(alert).toMatchSnapshot();
   expect(link).toBeInTheDocument();
   expect(link.href).toBe('http://localhost/#');
   expect(link.target).toBe('_blank');
   expect(link.querySelector('svg')).not.toBeUndefined();
 });
 
-test('renders header', () => {
-  const { queryByText, container } = render(<Alert header="Header" messages={[{ text: 'Success' }]} />);
+test('renders header', async () => {
+  render(<Alert header="Header" messages={[{ text: 'Success' }]} />);
+
+  const alert = await screen.findByRole('alert');
 
-  expect(container.firstChild).toMatchSnapshot();
-  expect(queryByText('Header')).not.toBeUndefined();
+  expect(alert).toMatchSnapshot();
+  expect(screen.getByRole('heading', { name: /header/i })).not.toBeUndefined();
 });
 
-test('renders close button', () => {
-  const { queryByRole, container } = render(<Alert onClose={() => null} messages={[{ text: 'Success' }]} />);
+test('renders close button', async () => {
+  render(<Alert onClose={() => null} messages={[{ text: 'Success' }]} />);
+
+  const alert = await screen.findByRole('alert');
 
-  expect(container.firstChild).toMatchSnapshot();
-  expect(queryByRole('button')).not.toBeUndefined();
+  expect(alert).toMatchSnapshot();
+  expect(screen.getByRole('button')).not.toBeUndefined();
 });
 
 test('trigger onClose', () => {
   const fn = jest.fn();
-  const { queryByRole } = render(<Alert onClose={fn} messages={[{ text: 'Success' }]} />);
+  render(<Alert onClose={fn} messages={[{ text: 'Success' }]} />);
 
-  const button = queryByRole('button') as HTMLButtonElement;
+  const button = screen.getByRole('button') as HTMLButtonElement;
 
   fireEvent.click(button);
 
   expect(fn).toHaveBeenCalled();
 });
 
-test('does not forward styles', () => {
-  const { container } = render(
-    <Alert messages={[{ text: 'Success' }]} className="test" style={{ background: 'red' }} />,
-  );
+test('does not forward styles', async () => {
+  render(<Alert messages={[{ text: 'Success' }]} className="test" style={{ background: 'red' }} />);
+
+  const alert = await screen.findByRole('alert');
+  const elem = screen.queryByText('Success');
 
-  expect(container.getElementsByClassName('test').length).toBe(0);
-  expect(container.firstChild).not.toHaveStyle('background: red');
+  expect(elem).not.toHaveClass('test');
```

Comment:
Why are we expecting the text `Success` to not have a class here, instead of the alert role?

Even so, I believe this is changing the behavior of the test. The diff from before is testing to see if any elements in the alert role does not contain a class `test`. While this is now testing if just the text `Success` doesn't contain the class `test`. For this test, this expect is a valid use-case for using `container`, based on the intentions of the test. (The `expect(alert).not.toHaveStyle...` is still good to keep.)

I believe this is what we might want 🤔 
```suggestion
  expect(container.getElementsByclassName('test')).toHaveLength(0);
```

Reply:
@chanceaclark,  I wonder when it’s acceptable to use container at all. I’ve gone through Kent's article again and it looks for me as he doesn’t recommend to use container in general.. Maybe you can advise some article or examples? Thanks.

Reply:
@bc-alexsaiannyi One of the primary goals of tests is to gain confidence in our application/page/component. RTL targets that goal  in a way to write tests that mimic how normal users would interact with it, which in turn, helps us gain confidence that our application/page/component works. With that in mind, Kent recommends avoiding using `container` because using it is how a developer would interact, not a normal user. However, for this test case we are trying to test how a developer interacts with it and gain confidence that our component doesn't pass in a `className` and `style` props. So I would say this is a perfectly acceptable usage of `container`. The bottom line is that we should ask ourselves, _is using `container` helping us gain confidence in our tests_? If yes, and we know what we are doing, go right ahead.

I would say the **TL'DR** is to avoid using `container` unless you are trying to test something around how a developer would interact. And it's really only in niche cases, like this `className` stuff since we don't want developers applying custom styles to some of our components.

Reply:
Thanks for explaining.